File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = [0] * len(nums)

        for i, val in enumerate(nums):
            if i == 0:
                dp[i] = val
            elif i == 1:
                dp[i] = max(val, dp[i - 1])
            else:
                dp[i] = max(val + dp[i - 2], dp[i - 1])
            
        return dp[-1]

File: Data Structures & Algorithms/target-sum/submission-1.cpp
class Solution {
    std::vector<std::vector<int>> dp;
    int totalSum;

public:
    int findTargetSumWays(vector<int>& nums, int target) {
        totalSum = std::accumulate(nums.begin(), nums.end(), 0);
        dp.assign(nums.size(), std::vector<int>(totalSum * 2 + 1, INT_MIN));

        return dfs(0, 0, target, nums);
    }

    int dfs(int index, int currSum, int target, std::vector<int>& nums) {
        if (index >= nums.size()) {
            return currSum == target ? 1 : 0;
        }

        if (dp[index][currSum + totalSum] != INT_MIN) {
            return dp[index][currSum + totalSum];
        }

        dp[index][currSum + totalSum] = dfs(index + 1, currSum + nums[index], target, nums) + dfs(index + 1, currSum - nums[index], target, nums);

        return dp[index][currSum + totalSum];
    }



};
